File: 437-path-sum-iii/path-sum-iii.c
#include <stdio.h>
#include <stdlib.h>

// Definition for a binary tree node.

int countPathsWithSum(struct TreeNode* node, long long currentSum, int targetSum) {
    if (node == NULL) {
        return 0;
    }

    currentSum += node->val;
    int totalPaths = 0;

    if (currentSum == targetSum) {
        totalPaths++;
    }

    totalPaths += countPathsWithSum(node->left, currentSum, targetSum);
    totalPaths += countPathsWithSum(node->right, currentSum, targetSum);

    return totalPaths;
}

int pathSum(struct TreeNode* root, int targetSum) {
    if (root == NULL) {
        return 0;
    }

    // Count paths starting from the current node
    int pathsFromRoot = countPathsWithSum(root, 0, targetSum);

    // Recursively try to find paths starting from the left and right child
    int pathsOnLeft = pathSum(root->left, targetSum);
    int pathsOnRight = pathSum(root->right, targetSum);

    return pathsFromRoot + pathsOnLeft + pathsOnRight;
}

File: 437-path-sum-iii/path-sum-iii.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int pathSum(TreeNode root, int targetSum) {
        if (root == null) return 0;
        Map<Long, Integer> prefixSum = new HashMap<>();
        prefixSum.put(0L, 1);
        return pathSumHelper(root, 0, targetSum, prefixSum);
    }
    
    private int pathSumHelper(TreeNode node, long currentSum, int targetSum, Map<Long, Integer> prefixSum) {
        if (node == null) return 0;
        
        currentSum += node.val;
        int count = prefixSum.getOrDefault(currentSum - targetSum, 0);
        
        prefixSum.put(currentSum, prefixSum.getOrDefault(currentSum, 0) + 1);
        
        count += pathSumHelper(node.left, currentSum, targetSum, prefixSum);
        count += pathSumHelper(node.right, currentSum, targetSum, prefixSum);
        
        prefixSum.put(currentSum, prefixSum.get(currentSum) - 1);
        
        return count;
    }
}